File: web-sites-hub/frontend-docs/banana-algo-core-docs/04-recursion-backtracking/backtracking.py
"""
递归与回溯算法实现
"""

# ========== 1. 全排列 ==========

def permute(nums):
    """全排列"""
    result = []
    
    def backtrack(path, choices):
        if len(path) == len(nums):
            result.append(path[:])
            return
        
        for i in range(len(choices)):
            path.append(choices[i])
            backtrack(path, choices[:i] + choices[i+1:])
            path.pop()
    
    backtrack([], nums)
    return result


# ========== 2. 组合 ==========

def combine(n, k):
    """从1到n选k个数的所有组合"""
    result = []
    
    def backtrack(start, path):
        if len(path) == k:
            result.append(path[:])
            return
        
        for i in range(start, n + 1):
            path.append(i)
            backtrack(i + 1, path)
            path.pop()
    
    backtrack(1, [])
    return result


# ========== 3. 子集 ==========

def subsets(nums):
    """所有子集"""
    result = []
    
    def backtrack(start, path):
        result.append(path[:])
        
        for i in range(start, len(nums)):
            path.append(nums[i])
            backtrack(i + 1, path)
            path.pop()
    
    backtrack(0, [])
    return result


# ========== 4. N皇后 ==========

def solve_n_queens(n):
    """N皇后问题"""
    result = []
    board = [['.'] * n for _ in range(n)]
    
    def is_valid(row, col):
        # 检查列
        for i in range(row):
            if board[i][col] == 'Q':
                return False
        
        # 检查左上对角线
        i, j = row - 1, col - 1
        while i >= 0 and j >= 0:
            if board[i][j] == 'Q':
                return False
            i -= 1
            j -= 1
        
        # 检查右上对角线
        i, j = row - 1, col + 1
        while i >= 0 and j < n:
            if board[i][j] == 'Q':
                return False
            i -= 1
            j += 1
        
        return True
    
    def backtrack(row):
        if row == n:
            result.append([''.join(row) for row in board])
            return
        
        for col in range(n):
            if is_valid(row, col):
                board[row][col] = 'Q'
                backtrack(row + 1)
                board[row][col] = '.'
    
    backtrack(0)
    return result


# ========== 5. 括号生成 ==========

def generate_parenthesis(n):
    """生成n对有效括号"""
    result = []
    
    def backtrack(path, left, right):
        if len(path) == 2 * n:
            result.append(path)
            return
        
        if left < n:
            backtrack(path + '(', left + 1, right)
        if right < left:
            backtrack(path + ')', left, right + 1)
    
    backtrack('', 0, 0)
    return result


# ========== 6. 组合总和 ==========

def combination_sum(candidates, target):
    """数字可重复使用"""
    result = []
    
    def backtrack(start, path, total):
        if total == target:
            result.append(path[:])
            return
        if total > target:
            return
        
        for i in range(start, len(candidates)):
            path.append(candidates[i])
            backtrack(i, path, total + candidates[i])  # i不是i+1，可重复
            path.pop()
    
    backtrack(0, [], 0)
    return result


# ========== 7. 单词搜索 ==========

def exist(board, word):
    """在网格中查找单词"""
    m, n = len(board), len(board[0])
    
    def backtrack(i, j, k):
        if k == len(word):
            return True
        
        if i < 0 or i >= m or j < 0 or j >= n or board[i][j] != word[k]:
            return False
        
        temp = board[i][j]
        board[i][j] = '#'  # 标记已访问
        
        found = (backtrack(i+1, j, k+1) or
                backtrack(i-1, j, k+1) or
                backtrack(i, j+1, k+1) or
                backtrack(i, j-1, k+1))
        
        board[i][j] = temp  # 恢复
        return found
    
    for i in range(m):
        for j in range(n):
            if backtrack(i, j, 0):
                return True
    return False


def demo():
    """演示回溯算法"""
    print("=== 回溯算法演示 ===\n")
    
    # 全排列
    print("全排列 [1,2,3]:")
    print(permute([1, 2, 3]))
    print()
    
    # 组合
    print("C(4,2) 组合:")
    print(combine(4, 2))
    print()
    
    # 子集
    print("子集 [1,2,3]:")
    print(subsets([1, 2, 3]))
    print()
    
    # N皇后
    print("4皇后问题:")
    solutions = solve_n_queens(4)
    print(f"共{len(solutions)}种解法")
    for i, solution in enumerate(solutions):
        print(f"\n解法{i+1}:")
        for row in solution:
            print(row)
    print()
    
    # 括号生成
    print("3对括号:")
    print(generate_parenthesis(3))
    print()
    
    # 组合总和
    print("组合总和 candidates=[2,3,6,7], target=7:")
    print(combination_sum([2, 3, 6, 7], 7))


if __name__ == '__main__':
    demo()

File: web-sites-hub/frontend-docs/banana-algo-core-docs/04-recursion-backtracking/backtracking.cpp
/**
 * 递归与回溯算法实现
 */

#include <iostream>
#include <vector>
#include <string>

using namespace std;

// ========== 全排列 ==========

void permuteHelper(vector<int>& nums, vector<bool>& used, 
                   vector<int>& path, vector<vector<int>>& result) {
    if (path.size() == nums.size()) {
        result.push_back(path);
        return;
    }
    
    for (int i = 0; i < nums.size(); i++) {
        if (used[i]) continue;
        
        used[i] = true;
        path.push_back(nums[i]);
        permuteHelper(nums, used, path, result);
        path.pop_back();
        used[i] = false;
    }
}

vector<vector<int>> permute(vector<int>& nums) {
    vector<vector<int>> result;
    vector<int> path;
    vector<bool> used(nums.size(), false);
    permuteHelper(nums, used, path, result);
    return result;
}

// ========== 组合 ==========

void combineHelper(int n, int k, int start, vector<int>& path, 
                   vector<vector<int>>& result) {
    if (path.size() == k) {
        result.push_back(path);
        return;
    }
    
    for (int i = start; i <= n; i++) {
        path.push_back(i);
        combineHelper(n, k, i + 1, path, result);
        path.pop_back();
    }
}

vector<vector<int>> combine(int n, int k) {
    vector<vector<int>> result;
    vector<int> path;
    combineHelper(n, k, 1, path, result);
    return result;
}

// ========== 子集 ==========

void subsetsHelper(vector<int>& nums, int start, vector<int>& path,
                   vector<vector<int>>& result) {
    result.push_back(path);
    
    for (int i = start; i < nums.size(); i++) {
        path.push_back(nums[i]);
        subsetsHelper(nums, i + 1, path, result);
        path.pop_back();
    }
}

vector<vector<int>> subsets(vector<int>& nums) {
    vector<vector<int>> result;
    vector<int> path;
    subsetsHelper(nums, 0, path, result);
    return result;
}

// ========== N皇后 ==========

bool isValid(const vector<string>& board, int row, int col, int n) {
    // 检查列
    for (int i = 0; i < row; i++) {
        if (board[i][col] == 'Q') return false;
    }
    
    // 检查左上对角线
    for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--) {
        if (board[i][j] == 'Q') return false;
    }
    
    // 检查右上对角线
    for (int i = row - 1, j = col + 1; i >= 0 && j < n; i--, j++) {
        if (board[i][j] == 'Q') return false;
    }
    
    return true;
}

void solveNQueensHelper(int n, int row, vector<string>& board,
                        vector<vector<string>>& result) {
    if (row == n) {
        result.push_back(board);
        return;
    }
    
    for (int col = 0; col < n; col++) {
        if (isValid(board, row, col, n)) {
            board[row][col] = 'Q';
            solveNQueensHelper(n, row + 1, board, result);
            board[row][col] = '.';
        }
    }
}

vector<vector<string>> solveNQueens(int n) {
    vector<vector<string>> result;
    vector<string> board(n, string(n, '.'));
    solveNQueensHelper(n, 0, board, result);
    return result;
}

// ========== 括号生成 ==========

void generateParenthesisHelper(int n, int left, int right,
                               string& path, vector<string>& result) {
    if (path.length() == 2 * n) {
        result.push_back(path);
        return;
    }
    
    if (left < n) {
        path.push_back('(');
        generateParenthesisHelper(n, left + 1, right, path, result);
        path.pop_back();
    }
    
    if (right < left) {
        path.push_back(')');
        generateParenthesisHelper(n, left, right + 1, path, result);
        path.pop_back();
    }
}

vector<string> generateParenthesis(int n) {
    vector<string> result;
    string path;
    generateParenthesisHelper(n, 0, 0, path, result);
    return result;
}

// 打印结果
template<typename T>
void printVector(const vector<T>& vec) {
    cout << "[";
    for (size_t i = 0; i < vec.size(); i++) {
        cout << vec[i];
        if (i < vec.size() - 1) cout << ", ";
    }
    cout << "]";
}

int main() {
    cout << "=== 回溯算法演示 ===" << endl << endl;
    
    // 全排列
    vector<int> nums = {1, 2, 3};
    cout << "全排列 [1,2,3]:" << endl;
    vector<vector<int>> perms = permute(nums);
    for (const auto& perm : perms) {
        cout << "  ";
        printVector(perm);
        cout << endl;
    }
    cout << endl;
    
    // 组合
    cout << "C(4,2) 组合:" << endl;
    vector<vector<int>> combs = combine(4, 2);
    for (const auto& comb : combs) {
        cout << "  ";
        printVector(comb);
        cout << endl;
    }
    cout << endl;
    
    // 子集
    vector<int> nums2 = {1, 2, 3};
    cout << "子集 [1,2,3]:" << endl;
    vector<vector<int>> subs = subsets(nums2);
    cout << "  共" << subs.size() << "个子集" << endl << endl;
    
    // N皇后
    cout << "4皇后问题:" << endl;
    vector<vector<string>> queens = solveNQueens(4);
    cout << "  共" << queens.size() << "种解法" << endl << endl;
    
    // 括号生成
    cout << "3对括号:" << endl;
    vector<string> parens = generateParenthesis(3);
    for (const auto& p : parens) {
        cout << "  " << p << endl;
    }
    
    return 0;
}

